examples/blob2d-outerloop/blob2d.cxx

```diff
@@ -12,10 +12,47 @@
 
 #include "bout/single_index_ops.hxx" // Operators at a single index
 
-#define RUN_WITH_RAJA 0
+#define RUN_WITH_RAJA 1
 
 #if BOUT_HAS_RAJA && RUN_WITH_RAJA
 #include "RAJA/RAJA.hpp" // using RAJA lib
+
+struct RajaWrapper {
+  RajaWrapper(const Region<Ind3D>& region) {
+    auto indices = region.getIndices(); // A std::vector of Ind3D objects
+    _ob_i_ind.reallocate(indices.size());
+    // Copy indices into Array
+    for(auto i = 0; i < indices.size(); i++) {
+      _ob_i_ind[i] = indices[i].ind;
+    }
+  }
+  template<typename F>
+  void operator<<(F f) {
+    // Get the raw pointer to use on the device
+    // Note: must be a local variable
+    int* _ob_i_ind_raw = &_ob_i_ind[0];
+    RAJA::forall<EXEC_POL>(RAJA::RangeSegment(0, _ob_i_ind.size()),
+			   [=] RAJA_DEVICE(int id) {
+			     // Look up index and call user function
+			     f(_ob_i_ind_raw[id]);
+			   });
+  }
+  // Note: This is private, but keyword not used due to CUDA limitation
```

Comment:
I've had a (very) quick attempt at reproducing this in plain CUDA and haven't managed it. I'd quite like to get to the bottom of this before it gets into `next` -- is it a particular version of CUDA, RAJA, or some interaction with BOUT++?

Reply:
It is odd. The error is `The enclosing parent function ("rhs") for an extended __device__ lambda cannot have private or protected access within its class`. It's something to do with "extended" (device) lambdas in CUDA:
https://docs.nvidia.com/cuda/cuda-c-programming-guide/#extended-lambda
There's a comment here which linked to the docs https://github.com/NVIDIA/thrust/issues/726#issuecomment-168733162

Reply:
It seems that the class *can* have private members, it's just that the function where the lambda function is defined can't be private or protected.